### PR TITLE
hashlink: init at 1.12

### LINKED ISSUE #168248: Hashlink - virtual machine for Haxe development
**Project description**
Hashlink is a virtual machine used in Haxe/Heaps game projects. This is the same stack that shiro games and motion twin uses to make things like Dead Cells and Northgard. The full steps for installing heaps (which goes through dependencies) can be found [here](https://heaps.io/documentation/installation.html), and one of the steps is to download hashlink and put it on the path.

**Metadata**

* homepage URL: https://hashlink.haxe.org/
* source URL: https://github.com/HaxeFoundation/hashlink/releases
* license: mit 
* platforms: linux, mac, windows

### PR DESCRIPTION
**Pull request by locallycompact**
###### Description of changes

Closes #168289
Closes #168248

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/hashlink/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  version = "1.12";
+
+  src = fetchFromGitHub {
+    owner = "HaxeFoundation";
+    repo = "hashlink";
+    rev = version;
+    sha256 = "AiUGhTxz4Pkrks4oE+SAuAQPMuC5T2B6jo3Jd3sNrkQ=";
+  };
+
+  buildInputs = [ libpng libjpeg_turbo libvorbis openal SDL2 mbedtls libuv pcre libGL libGLU sqlite ];
```
Comment (superherointj) [inline pkgs/development/interpreters/hashlink/default.nix:28]: Suggestion of formatting:
1) list attributes in phases order (fetch => patch => configure => build => check => install)
2) list values in alphabetical order.
3) when a list has more than 3 items, list one item per line.

Comment (locallycompact) [reply pkgs/development/interpreters/hashlink/default.nix:28]: Are attributes not in phase order here?

**Comment by robinp** [comment]
FYI I tried this on a non-nixos machine as well - hashlink there starts at least (doesn't exit with unable to load hdll), but exits with couldn't create window. strace indicates it couldn't find some GLX libs. So there might be some setups still where this binary doesn't work as-is, but probably that's something to sort later.

(Tried to get mesa built with hashlink if that would help, but that seems non-trivial, at least not just a simple make & go)

File: pkgs/development/interpreters/hashlink/default.nix (hunk 1 above)
Comment (superherointj) [reply pkgs/development/interpreters/hashlink/default.nix]: `patches` is patchPhase.
`makeFlags` is configurePhase.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -6903,6 +6903,8 @@ with pkgs;
 
   httplab = callPackage ../tools/networking/httplab { };
 
+  hashlink = callPackage ../development/interpreters/hashlink { };
```
Comment (superherointj) [inline pkgs/top-level/all-packages.nix:6906]: `hashlink` is not at alphabetical order.
Also add it to the correct section ( `development/interpreters`)

File: pkgs/development/interpreters/hashlink/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  meta = with lib; {
+    description = "A virtual machine for Haxe";
+    homepage = "https://hashlink.haxe.org/";
+    license = licenses.mit;
+    platforms = platforms.unix;
+    maintainers = with maintainers; [ iblech locallycompact ];
+  };
+
+  patches = [ ./hashlink.patch ];
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/hashlink/default.nix:52]:
```suggestion
```

File: pkgs/development/interpreters/hashlink/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+stdenv.mkDerivation rec {
+  pname = "hashlink";
+  version = "1.12";
+
+  src = fetchFromGitHub {
+    owner = "HaxeFoundation";
+    repo = "hashlink";
+    rev = version;
+    sha256 = "AiUGhTxz4Pkrks4oE+SAuAQPMuC5T2B6jo3Jd3sNrkQ=";
+  };
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/hashlink/default.nix:26]:
```suggestion
  };

  patches = [ ./hashlink.patch ];
```

**Comment by SuperSandro2000** [comment]
Either we mark the package broken on aarch64 or remove the x86 only flags:


```
       > configuring
       > no configure script, doing nothing
       > building
       > build flags: SHELL=/nix/store/52wfs5n7vx8yyq5d7l0r734p78hn0wmd-bash-5.1-p16/bin/bash PREFIX=\$\(out\)
       > gcc -Wall -O3 -I src -msse2 -mfpmath=sse -std=c11 -D LIBHL_EXPORTS -m64 -fPIC -pthread -fno-omit-frame-pointer -o include/pcre/pcre_chartables.o -c include/pcre/pcre_chartables.c
       > gcc: error: unrecognized command line option '-msse2'
       > gcc: error: unrecognized command line option '-mfpmath=sse'
       > gcc: error: unrecognized command line option '-m64'
       > make: *** [Makefile:143: include/pcre/pcre_chartables.o] Error 1
```

**Comment by locallycompact** [comment]
Ok. I don't know what that is.

**Comment by SuperSandro2000** [comment]
The build error on aarch64-linux which ofborg build.

**Comment by locallycompact** [comment]
Do I need to do something about that?

**Comment by SuperSandro2000** [comment]
> Do I need to do something about that?

Yes. Either we fix it by removing the not supported flags or we mark it broken on that platform. Since upstream does not seem to have support for aarch64 I would suggest to mark it broken.

**Comment by locallycompact** [comment]
Is it ok like this?

File: pkgs/development/interpreters/hashlink/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    libuv
+    libvorbis
+    mbedtls
+    openal
+    pcre
+    SDL2
+    sqlite
+  ];
+
+  makeFlags = [ "PREFIX=$(out)" ];
```
Comment (superherointj) [inline pkgs/development/interpreters/hashlink/default.nix:44]:
```suggestion
  makeFlags = [ "PREFIX=$(out)" ];
```
makeFlags is `configurePhase` which comes before `buildPhase`.

Comment (locallycompact) [reply pkgs/development/interpreters/hashlink/default.nix]: How about now?

**Comment by tobil4sk** [comment]
For future reference: the patch that is a part of the installation process for this package has been applied upstream:
https://github.com/HaxeFoundation/hashlink/commit/d02bbf6cbc977e1c86e81d3024e5094ea09a94b0

Therefore, the manual patch can be removed for hashlink 1.13 and future versions.

**Pull request merged**